### ISSUE TITLE
Upstream Proxy + Caching Images

### ISSUE THREAD

**Issue by shirshak55** [httptoolkit/mockttp#90]
I am currently trying to cache images, and send the request to upstream httptoolkit servers. The goal is to not see useless request in httptoolkit app.

```
 let extensions = ['.css', '.js', '.png', '.gif', '.ico'];
  server
    .forAnyRequest()
    .matching(async (req) => {
      if (extensions.some((v) => req.url.includes(v))) {
        return true;
      }
      return false;
    })
    .thenPassThrough({
      beforeResponse: async (response) => {
        let link = ids_to_link[response.id];
        let hash = Buffer.from(ids_to_link[response.id]).toString('base64');
        let pth = path.join(`./cache/${hash}`);

        if (hash in dataStore) {
          return await fsp.readFile(pth);
        } else {
          console.log('Storing in cache', link);
          dataStore[hash] = pth;
          await fsp.writeFile(pth, await response.body.getDecodedBuffer());
        }

        return {};
      },
    });
  server.forUnmatchedRequest().thenPassThrough({
    trustAdditionalCAs: [
      { certPath: '/home/quantum/.config/httptoolkit/ca.pem' },
    ],

    proxyConfig: {
      proxyUrl: 'http://127.0.0.1:8000',
    },
  });
```

Obviously, it doesn't work. Let me know if there is a way. Thanks

**Comment by pimterry** [comment]
> Obviously, it doesn't work

What exactly doesn't work? Is there an error, is there any output, what's the behaviour you see?

A few things I can see from a quick skim:
* You're only using `beforeResponse` (not `beforeRequest`) so the request will always be sent upstream (`beforeResponse` only fires after the response has already been received).
* In `beforeResponse`, you're returning the raw body data directly, not an object with fields, e.g. `{ status: 200, headers: { ... }, body: yourCachedBody }`.
* You probably want to add `.always()` to that rule, so that it runs forever. In this case, because it's the only rule except the unmatched requests rule, it won't matter, but other this might do confusing things if you add more rules later.

I think you want to change to:
* Use a `beforeRequest` callback to decide whether to forward the request or return a response from your cache
* In that callback, either return undefined (to forward the request as normal) or `{ response: { status: 200, headers: cachedHeaders, body: cachedBody } }` to inject a response to taht request using your data (this will skipping forwarding, and just reply from the cache immediately). The full docs for this return type are here: https://httptoolkit.github.io/mockttp/interfaces/requestHandlers.CallbackRequestResult.html.
* Use a `beforeResponse` callback too, but just to cache the real responses. Note that `beforeResponse` is _not_ called if `beforeRequest` overrides everything and returns a response for itself instead - it's only called for real upstream responses.

That should do what you're looking for. Does that make sense?

Alternatively, there is a different approach, where you create two rules, separately handling cached/uncached content, something like this: 

* `forAnyRequest().matching(req => isInCache(req)).always().thenCallback((req) => buildResponseFromCache(req))`
* `forAnyRequest().matching(req => !isInCache(req) && hasCacheableExtension(req)).always().thenPassThrough({ ..., beforeResponse: (res) => { cacheResponse(res) } })`

**Comment by pimterry** [comment]
> It would be nice if we can access the request URL inside beforeResponse

Yes! I'm very open to this, if you want to open a PR to pass a second request argument to beforeResponse. If you do do that, make sure to handle that parameter in the serialization logic, and add a remote-client test to check that the result works, since it's a bit tricky.

> If multiple request matches forAnyRequest it seems to go in serial order and after it finishes mockttp uses the last handler. It's very confusing?

This is what I mean by using `.always()` above. By default, the rules run in order, until the last rule, which repeats forever. When writing tests, this is usually what you want and it's very convenient for defining a series of responses. In this case though, you probably want to use `.always()` (or other options like `.once()`, `.twice`, `.times(n)`) to control exactly how many times each rule runs explicitly instead.

> It would be nice if there were some script generator that would generate via HTTP Toolkit from Mock Tab

It would! I think that's possible (the rules data is very closely related between the two) but I really don't have much time at the moment for this. Definitely something I'm planning for the future though. I'd also love to add "custom script" rules in HTTP Toolkit, where you can just write Mockttp logic in there directly. One day :smile:
